### PR TITLE
Properly Record GRPC Error

### PR DESCRIPTION
**Pull request by pongad**
If a request's context times out or is cancelled when we are sleeping
between retries,
we previously return the context's error.

This patch returns both the context's error and the GRPC call's,
by wrapping them both in InvokeError.

### CODE REVIEW

**Comment by pongad** [comment]
CC @okdave @garye

File: invoke.go [hunk 1]
```diff
@@ -33,37 +37,60 @@ func invokeWithRetry(ctx context.Context, stub APICall, callSettings CallSetting
...
 		return invokeWithRetry(ctx, stub, *settings)
 	}
-	return stub(ctx)
+	if err := stub(ctx); err != nil {
+		return invokeError{grpcErr: err}
+	}
+	return nil
+}
+
+// InvokeError records the GRPC error from the last completed GRPC call.
```
Comment (okdave) [inline invoke.go]: Do you need an interface type here? Can the function just return a concrete &InvokeError (where InvokeError is a struct)?

File: invoke.go [hunk 2]
```diff
@@ -11,10 +12,13 @@ import (
 // A user defined call stub.
 type APICall func(context.Context) error
 
-// scaleDuration returns the product of a and mult.
-func scaleDuration(a time.Duration, mult float64) time.Duration {
-	ns := float64(a) * mult
-	return time.Duration(ns)
+// scaleDuration returns the duration in `mult` after `d`
```
Comment (okdave) [inline invoke.go]: This seems like an unrelated change.
Also, don't use markdown in the comments.

// scaleDuration scales the duration d by mult.

File: invoke.go (hunk 1 above)
Comment (pongad) [inline invoke.go]: Good point. I was thinking of documenting the interface so people will use it instead of accessing the member, but the members are private anyway.

File: invoke.go (hunk 2 above)
Comment (pongad) [inline invoke.go]: This change was made so that we don't need to create 3 new contexts every time we want to sleep at https://github.com/googleapis/gax-go/pull/35/files#diff-39e0330b64fba4170680bf6bd806b22fL51 .

Since I need to rewrite a lot of invokeWithRetry anyway (this function isn't used elsewhere), I thought I might as well simplify it. Would you like me to separate this into another change?

**Comment by pongad** [comment]
@okdave I thought this would be a simple issue I could work on while waiting for discussions around the retry improvement proposal to resolve. It sounds like it's turning out to be a lot more subtle than I thought. Undecided details in the proposal will probably affect implementation and docs. I feel tempted to close this pull request for the time being and reopen when dust settles on the proposal. Do you think that will make things cleaner?
